File: src/Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int nodenum = 0;
        ListNode *p = head;
        while(p != NULL)
        {
            nodenum++;
            p = p->next;
        }
        if(nodenum<2)
            return head;
        p = head;
        ListNode *q = p->next;
        head = q;
        p->next = q->next;
        q->next = p;
        int n = nodenum / 2;
        while(n>1)
        {
            p = p->next;
            q = q->next;
            q->next = p->next;
            q = p->next;
            p->next = q->next;
            q->next = p;
            n--;
        }
        return head;        
    }
};

File: python/Swap Nodes in Pairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        if head is None or head.next is None:
            return head

        retHead = head.next
        p = head
        q = head.next
        while True:
            p.next = q.next
            q.next = p
            p = p.next
            if p is None or p.next is None:
                break
            if p.next != None:
                q.next.next = p.next
                q = p.next

        return retHead